src/db/util/conditions.ts

```diff
@@ -0,0 +1,178 @@
+/* eslint-disable @typescript-eslint/no-namespace */
+// we use namespaces here to neatly organize and scope the many type and type guard definitions
+
+import Knex = require('knex');
+
+/**
+ * Symbols to use for where condition construction
+ */
+export namespace ConditionSymbols {
+  export const BUILDER = Symbol('builder');
+  export const AND = Symbol('and');
+  export const OR = Symbol('or');
+
+  /**
+   * Symbols to use for where condition construction
+   */
+  export const Cond = {
+    BUILDER: BUILDER,
+    AND: AND,
+    OR: OR,
+  } as const;
+}
+
+export const Cond = ConditionSymbols.Cond;
+
+/**
+ * Symbols to use when constructing conditions for a single property
+ */
+export namespace PropertySymbols {
+  export const IN = Symbol('in');
+  export const NOT_IN = Symbol('not in');
+
+  /**
+   * Symbols to use when constructing conditions for a single property
+   */
+  export const Op = {
+    IN: IN,
+    NOT_IN: NOT_IN,
+  } as const;
+}
+
+export const Op = PropertySymbols.Op;
+
+namespace PropertyConditions {
+  export type EqualityCondition<T> = T;
+  export type InCondition<T> = {
+    [Op.IN]: Iterable<T>;
+  };
+  export type NotInCondition<T> = {
+    [Op.NOT_IN]: Iterable<T>;
+  };
+  /**
+   * A condition that must hold over a single property whose type is T
+   */
+  export type Condition<T> =
+    | EqualityCondition<T>
+    | InCondition<T>
+    | NotInCondition<T>;
+
+  export const isEqualityCondition = <T>(
+    condition: Condition<T>
+  ): condition is EqualityCondition<T> =>
+    typeof condition !== 'object' || condition === null;
```

Comment:
I wanted to break this by supplying `undefined` (literal) and see what happens. Since the condition resolves to `true`, `undefined` is supplied to `knex`'s `where` and we get "Undefined binding(s) detected when compiling WHERE" error.

I don't know what to make out of this though. I don't think sequelize failed in this case (not sure), but I don't want to silently convert it to `null` check either. No one will supply `undefined` literal (hopefully), but maybe it comes from a variable. Any thoughts on this?

Reply:
I dont think this is something that we can easuly make a type error due to the "Partial" nature of the condition over properties, as long as something is defined as `[...]?: T`, then i think you can assign undefined to it. If that's the case, then I think we probably just want to throw an error. If we're not happy with it being given to `knex` to catch, we can try and catch it earlier, which would happen if we just modify this condition slightly